### ISSUE TITLE
HNC: webhooks broken in K8s 1.19 due to Golang 1.15

### ISSUE THREAD

**Issue by adrianludwin** [kubernetes-sigs/multi-tenancy#1100]
cc @mbtamuli. See https://github.com/kubernetes-sigs/multi-tenancy/issues/1096#issuecomment-690868907.

/assign @yiqigao217 

{"level":"info","ts":1599798100.9179494,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-hnc-x-k8s-io-v1alpha2-hierarchyconfigurations"}
{"level":"info","ts":1599798100.9179764,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-objects"}
{"level":"info","ts":1599798100.9179943,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-hnc-x-k8s-io-v1alpha2-hncconfigurations"}
{"level":"info","ts":1599798100.9180129,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-hnc-x-k8s-io-v1alpha2-subnamespaceanchors"}
{"level":"info","ts":1599798100.9180274,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/validate-v1-namespace"}
{"level":"info","ts":1599798100.9180343,"logger":"controller-runtime.builder","msg":"skip registering a mutating webhook, admission.Defaulter interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=HNCConfiguration"}
{"level":"info","ts":1599798100.91805,"logger":"controller-runtime.builder","msg":"skip registering a validating webhook, admission.Validator interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=HNCConfiguration"}
{"level":"info","ts":1599798100.9180915,"logger":"controller-runtime.webhook","msg":"registering webhook","path":"/convert"}
{"level":"info","ts":1599798100.9180963,"logger":"controller-runtime.builder","msg":"conversion webhook enabled","object":{"name":""}}
{"level":"info","ts":1599798100.9181077,"logger":"controller-runtime.builder","msg":"skip registering a mutating webhook, admission.Defaulter interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=HierarchyConfiguration"}
{"level":"info","ts":1599798100.918112,"logger":"controller-runtime.builder","msg":"skip registering a validating webhook, admission.Validator interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=HierarchyConfiguration"}
{"level":"info","ts":1599798100.9181347,"logger":"controller-runtime.builder","msg":"conversion webhook enabled","object":{"name":""}}
{"level":"info","ts":1599798100.91815,"logger":"controller-runtime.builder","msg":"skip registering a mutating webhook, admission.Defaulter interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=SubnamespaceAnchor"}
{"level":"info","ts":1599798100.9181566,"logger":"controller-runtime.builder","msg":"skip registering a validating webhook, admission.Validator interface is not implemented","GVK":"hnc.x-k8s.io/v1alpha2, Kind=SubnamespaceAnchor"}
{"level":"info","ts":1599798100.9181774,"logger":"controller-runtime.builder","msg":"conversion webhook enabled","object":{"name":""}}

**Comment by mbtamuli** [comment]
I tried deploying the latest version using 
```
HNC_VERSION=v0.5.2
kubectl apply -f https://github.com/kubernetes-sigs/multi-tenancy/releases/download/hnc-${HNC_VERSION}/hnc-manager.yaml
```

Even then the logs from the manager container shows some errors. Logs - 
[manager.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5233474/manager.log)

**Comment by mbtamuli** [comment]
Could it be a version issue of Kind or Kubernetes?

I'm using 
```
kind version
kind v0.9.0 go1.15.2 linux/amd64

kubectl get nodes
NAME                 STATUS   ROLES    AGE     VERSION
kind-control-plane   Ready    master   6m40s   v1.19.1
kind-worker          Ready    <none>   6m4s    v1.19.1
kind-worker2         Ready    <none>   6m4s    v1.19.1
```

**Comment by adrianludwin** [comment]
Ah, all those "skip registering a mutating webhook" messages are harmless. They're all about defaults that are set on the API types themselves, but we don't use that feature. Our webhooks are registered _explicitly_ and aren't affected by these message. (This is all detailed `controller-runtime` stuff).

I think this is the same as https://github.com/open-policy-agent/gatekeeper/pull/811, and we need to port that here. That is: HNC is current broken on K8s 1.19, because that's built with Go 1.15 which changes the way we handle certificates. Updating the comment to better describe this.

/assign @adrianludwin

**Comment by adrianludwin** [comment]
We should fix this in the 0.5 branch. I'll do it.

**Comment by mbtamuli** [comment]
> That is: HNC is current broken on K8s 1.19, because that's built with Go 1.15 which changes the way we handle certificates. Updating the comment to better describe this.

Do you mean it should work with k8s <1.19?

**Comment by adrianludwin** [comment]
That's right - I've personally tested HNC up to 1.18 (on GKE) but not yet
on 1.19, I'll try to get to this by the end of the week.

**Comment by mbtamuli** [comment]
Hey @adrianludwin,

I was trying to run the tests. I am always getting errors no matter the kubernetes version I'm using.

I've tested with GKE - v1.15, v1.16 and v1.17, the ones you mentioned here - in [Testing Signoff](https://github.com/kubernetes-sigs/multi-tenancy/releases/tag/hnc-v0.5.2)

Here are the logs
[test-e2e-v115.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5243107/test-e2e-v115.log)
[test-e2e-v116.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5243108/test-e2e-v116.log)
[test-e2e-v117.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5243109/test-e2e-v117.log)

---
Am I running the e2e tests wrong? I was getting even errors without the `HNC_REPAIR` not set.

I did the following - 
```
gcloud container clusters create hnc-test-v117 --preemptible --num-nodes 1 --cluster-version 1.17 --release-channel rapid
HNC_REPAIR=true make test-e2e 2>&1 | tee test-e2e-v117.log

gcloud container clusters create hnc-test-v116 --preemptible --num-nodes 1 --cluster-version 1.16 --release-channel regular
HNC_REPAIR=true make test-e2e 2>&1 | tee test-e2e-v116.log

gcloud container clusters create hnc-test-v115 --preemptible --num-nodes 1 --cluster-version 1.15 --release-channel stable
HNC_REPAIR=true make test-e2e 2>&1 | tee test-e2e-v115.log
```

**Comment by adrianludwin** [comment]
@mbtamuli  - I just discovered that we're broken at HEAD. Did you try this on v0.5.2? See [here](https://github.com/kubernetes-sigs/controller-runtime/issues/1148#issuecomment-694455506) for details. We hadn't noticed this because we were testing on clusters that previously had HNC installed.

If you're not testing on HEAD, can you please attach the logs from HNC itself, and not just the tests? You can get the logs via `kubectl logs` or `make deploy-watch`.

**Comment by adrianludwin** [comment]
I'm actively working on this, but I likely will only be able to fix this on Monday.

**Comment by mbtamuli** [comment]
> Did you try this on v0.5.2?

Yes I did. Updated the previous comment with the deployment steps. Just followed the deployement steps mentioned in the release.

**Comment by adrianludwin** [comment]
Oh, you need to set `HNC_REPAIR` to `https://github.com/kubernetes-sigs/multi-tenancy/releases/download/hnc-${HNC_VERSION}/hnc-manager.yaml` (with the env var filled in). The test runs `kubectl apply -f ${HNC_REPAIR}` after it (deliberately) breaks the HNC installation.

If you leave the variable unset, it will skip some of the tests but it will run enough other things to be confident that you've at least installed it correctly.

Sorry, I should document that env var better.

**Comment by adrianludwin** [comment]
I don't know how to run Kind on Github Actions. We use Prow and we have some pointers on how to run Kind in Prow, so we'll likely be setting that up soon-ish.

**Comment by mbtamuli** [comment]
> Oh, you need to set HNC_REPAIR to https://github.com/kubernetes-sigs/multi-tenancy/releases/download/hnc-${HNC_VERSION}/hnc-manager.yaml (with the env var filled in). The test runs kubectl apply -f ${HNC_REPAIR} after it (deliberately) breaks the HNC installation.

Ah, I should've taken a better look at the Makefile or the existing e2e tests. Will try setting `HNC_REPAIR` to the manifest path.

---

> We use Prow and we have some pointers on how to run Kind in Prow, so we'll likely be setting that up soon-ish.

Can you mention one version of HNC(or commit), that you've tried with one version of KIND and one version of kubernetes, where the e2e tests finished successfully?
For example, you've tested successfully with
HNC - 0.5.2 AND Kind - 0.9.0 AND Kubernetes - 1.16

**Comment by mbtamuli** [comment]
Hey @adrianludwin, I got these logs. I tested these on GKE 1.17

```
gcloud container clusters create hnc-test-v117 --preemptible --num-nodes 1 --cluster-version 1.17 --release-channel rapid
```

**HNC - v.0.5.2**
 - [hnc-v117-0.5.2.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5256144/hnc-v117.log)
 - [test-e2e-v117-0.5.2.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5256145/test-e2e-v117.log)

**HNC - v.0.5.1**
- [hnc-v117-0.5.1.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5256152/hnc-v117.log)
- [test-e2e-v117-0.5.1.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5256153/test-e2e-v117.log)

**Comment by adrianludwin** [comment]
I think I saw the same thing recently and it was because I had the wrong version of `kubectl-hns`. Can you confirm you have the v0.5.x version of kubectl-hns on your path?

**Comment by mbtamuli** [comment]
Hey, I did try with the correct version of kubectl-hns. This time a few of the tests passed.

`FAIL! -- 20 Passed | 8 Failed | 0 Pending | 0 Skipped`

**HNC - v.0.5.1**
- [hnc-v117.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5262066/hnc-v117.log)
- [test-e2e-v117.log](https://github.com/kubernetes-sigs/multi-tenancy/files/5262067/test-e2e-v117.log)

**Comment by adrianludwin** [comment]
It seems like you're running some weird combination of branch (0.5) and
master (will be 0.6) based on this error message:

          s: "Error: exit status 1\nOutput: \nCould not update the HNC
Configuration: admission webhook \"hncconfigurations.hnc.x-k8s.io\" denied
the request: Unrecognized mode 'Ignore' for /v1, Kind=Secret\n",

The "Ignore" mode (upper-case) was introduced after 0.5, so 0.5 won't
recognize it. In 0.5, it was called "ignore" (lower-case). Are you running
the master version of the e2e tests? If so, make sure you switch to the
hnc-v0.5 branch and run the tests via "make e2e-test" (and not "make
test-e2e," as it was renamed after 0.5).

**Comment by adrianludwin** [comment]
Fixed in v0.5 by #1125, and will be fixed in master by #1128. Closing because it's fixed in v0.5 (we should release v0.5.3 shortly).

/close
